dotenv-linter/src/schema.rs

```diff
@@ -0,0 +1,484 @@
+use regex::Regex;
+use serde::Deserialize;
+use std::{collections::HashMap, fs::File, io::BufReader, path::Path};
+
+// the Readxxx structs are used to deserialize via serde
+// they are then post processed into the xxx structs
+
+#[derive(Deserialize, Default)]
+#[serde(default)]
+struct ReadDotEnvSchema {
+    pub version: String,
+    pub allow_other_keys: bool,
+    pub entries: Vec<ReadSchemaEntry>,
+}
+
+#[derive(Deserialize, Default)]
+#[serde(default)]
+pub struct ReadSchemaEntry {
+    pub key: String,
+    pub required: bool,
+    #[serde(rename = "type")]
+    pub value_type: SchemaValueType,
+    pub regex: Option<String>,
+}
+
+pub struct DotEnvSchema {
+    pub version: String,
+    pub allow_other_keys: bool,
+    pub entries: HashMap<String, SchemaEntry>,
+}
+
+pub struct SchemaEntry {
+    pub key: String,
+    pub required: bool,
+    pub value_type: SchemaValueType,
+    pub regex: Option<Regex>,
+}
+
+impl SchemaEntry {
+    pub fn new(read_entry: ReadSchemaEntry) -> Result<Self, regex::Error> {
+        let mut se = SchemaEntry {
+            key: read_entry.key,
+            required: read_entry.required,
+            value_type: read_entry.value_type,
+            regex: None,
+        };
+        if let Some(reg) = read_entry.regex {
+            se.regex = Some(Regex::new(&reg)?);
+        }
+        Ok(se)
+    }
+}
+
+#[derive(Deserialize, Default)]
+pub enum SchemaValueType {
+    #[default]
+    String,
+    Integer,
+    Float,
+    Boolean,
+    Url,
+    Email,
+}
+
+impl DotEnvSchema {
+    pub fn load(path: &Path) -> Result<Self, std::io::Error> {
+        let file = File::open(path)?;
+        let reader = BufReader::new(file);
+        let read_schema: ReadDotEnvSchema = serde_json::from_reader(reader)?;
```

Comment:
it does not work, it did originally but once the entries was changed to HashMap and the regex was compiled into the schema at load time it is no longer possible

Reply:
Yeah, I'm good with that. Actually I mentioned this way [earlier](https://github.com/dotenv-linter/dotenv-linter/pull/761#discussion_r1518575976):
> Also alternative to consider: instead of `array`, try to use `Object` in `JSON`, where `key` is env variable name 🤔 This can be deserialized into map

Any valid Unicode string is valid json key - so we should have no issues with that

Reply:
Another "unresolved" question regarding schema deserialization part is https://github.com/dotenv-linter/dotenv-linter/pull/761#issuecomment-1991243721 (also there is an example in rust-playground)

I tend to keep type-specific rules attached to the type itself.

Reply:
> Another "unresolved" question regarding schema deserialization part is [#761 (comment)](https://github.com/dotenv-linter/dotenv-linter/pull/761#issuecomment-1991243721) (also there is an example in rust-playground)
> 
> I tend to keep type-specific rules attached to the type itself (like `regex` belongs to `string`). Just to keep our code more extensible & clear for the future

I like your idea 👍